packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -2060,33 +2073,56 @@ function intoDeltaImpl(
 	change: FieldChangeMap,
 	nodeChanges: ChangeAtomIdBTree<NodeChangeset>,
 	fieldKinds: ReadonlyMap<FieldKindIdentifier, FieldKindWithEditor>,
-): Map<FieldKey, DeltaFieldChanges> {
+): [Map<FieldKey, DeltaFieldChanges>, DeltaDetachedNodeChanges[], DeltaDetachedNodeRename[]] {
 	const delta: Map<FieldKey, DeltaFieldChanges> = new Map();
+	const global: DeltaDetachedNodeChanges[] = [];
```

Comment:
If I'm following the code correctly, you're using an approach where the global and rename changes are copied upwards at every layer. For a changeset with N such changes in a single field at depth D, this approach will lead to O(N*D) operations.

This can be avoided by declaring a single pair of arrays for the whole changeset and passing a reference to them into recursive calls. It will also make the code simpler.

Reply:
This is how I initially had it and then decided on the current approach instead because it seems like we generally want to prefer code that doesn't have side effects. Is there a general rule for when we should prefer what? I guess when it's a perf hit?

Reply:
Good question. Aside from perf considerations, I would say that side effects are more of a problem when...
1. The contract ends up being a mix of side effects and returned values. This is the case here, and it makes me sad.
2. The side effects occur across a contract boundary where callers and callees may be implemented by different parties. This is not the case here, since both sides are parts of ModularChangeFamily's implementation details.

I can certainly appreciate the desire to avoid side-effect. The complexity cost seems pretty low here: the code with side effects ends up being slightly easier to follow (at least to my eyes). The contract for intoDeltaImpl still ends up more complicated though.